sched/scheduler/task_scheduler.go

```diff
@@ -39,21 +39,23 @@ func getTaskAssignments(cs *clusterState, jobs []*jobState,
 	}
 	defer stat.Latency(stats.SchedTaskAssignmentsLatency_ms).Time().Stop()
 
-	if config == nil {
-		config = &SchedulerConfig{
-			SoftMaxSchedulableTasks: DefaultSoftMaxSchedulableTasks,
-		}
-	}
-
 	// Exit if there are no unscheduled tasks.
+	totalOutstandingTasks := 0
 	totalUnschedTasks := 0
 	for _, j := range jobs {
+		totalOutstandingTasks += (len(j.Tasks) - j.TasksCompleted)
 		totalUnschedTasks += (len(j.Tasks) - j.TasksCompleted - j.TasksRunning)
 	}
 	if totalUnschedTasks == 0 {
 		return nil, nil
 	}
 
+	// Udate SoftMaxSchedulableTasks based on number of healthy nodes and the total number of tasks.
+	if config == nil {
+		config = &SchedulerConfig{}
+	}
+	config.SoftMaxSchedulableTasks = max(config.SoftMaxSchedulableTasks, totalOutstandingTasks, len(cs.nodes))
```

Comment:
I've been trying to understand the scheduling algorithm - need much more time to digest it, perhaps we can review it in tech discussion and update the write up.... given that:
- config.SoftMaxSchedulableTasks will be ever increasing since config.SoftMaxSchedulableTasks is part of the max statement - this seem suspicious

do we have tests validating the scheduling?